### PR TITLE
Re-add token scheme checks

### PR DESCRIPTION
**Pull request by thibault-martinez**
## Change checklist

- [x] I have followed the contribution guidelines for this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have checked that new and existing unit tests pass locally with my changes
- [x] I have updated the CHANGELOG.md, if my changes are significant enough

### CODE REVIEW

**Review by bingyanglin — DISMISSED** [review]

File: bee-message/src/output/foundry.rs [hunk 1]
```diff
@@ -366,13 +367,26 @@ impl StateTransitionVerifier for FoundryOutput {
             return Err(StateTransitionError::MissingAliasForFoundry);
         }
 
+        let token_id = next_state.token_id();
+        let output_tokens = context.output_native_tokens.get(&token_id).copied().unwrap_or_default();
+        let TokenScheme::Simple(ref next_token_scheme) = next_state.token_scheme;
+
+        // No native tokens should be referenced prior to the foundry creation.
+        if context.input_native_tokens.get(&token_id).is_some() {
```
Comment (Alex6323) [inline bee-message/src/output/foundry.rs]: Is there no `contains(&token_id)`?

File: bee-message/src/output/foundry.rs [hunk 2]
```diff
@@ -366,13 +367,26 @@ impl StateTransitionVerifier for FoundryOutput {
             return Err(StateTransitionError::MissingAliasForFoundry);
         }
 
+        let token_id = next_state.token_id();
+        let output_tokens = context.output_native_tokens.get(&token_id).copied().unwrap_or_default();
```
Comment (Alex6323) [inline bee-message/src/output/foundry.rs:371]: Would the default always lead to a `StateTransitionError::InconsistenNativeTokens` in line 379?

File: bee-message/src/output/foundry.rs [hunk 3]
```diff
@@ -383,17 +397,73 @@ impl StateTransitionVerifier for FoundryOutput {
...
 
+        match input_tokens.cmp(&output_tokens) {
+            Ordering::Less => {
+                // Minting
+
+                if &(current_token_scheme.minted_tokens() + (output_tokens - input_tokens))
+                    != next_token_scheme.minted_tokens()
+                    || current_token_scheme.melted_tokens() != next_token_scheme.melted_tokens()
+                {
+                    return Err(StateTransitionError::InconsistentNativeTokens);
```
Comment (Alex6323) [inline bee-message/src/output/foundry.rs]: Why do all three different state transitions create the same error? Wouldn't it make sense to have:
```rust
StateTransitionError::InconsistentMinting;
StateTransitionError::InconsistentTransitioning;
StateTransitionError::InconsistentMelting;
```
or something similar to make it easier to pinpoint what went wrong?

File: bee-message/src/output/foundry.rs [hunk 4]
```diff
@@ -383,17 +397,73 @@ impl StateTransitionVerifier for FoundryOutput {
...
         Ok(())
     }
 
-    fn destruction(_current_state: &Self, _context: &ValidationContext) -> Result<(), StateTransitionError> {
+    fn destruction(current_state: &Self, context: &ValidationContext) -> Result<(), StateTransitionError> {
+        let token_id = current_state.token_id();
+        let input_tokens = context.input_native_tokens.get(&token_id).copied().unwrap_or_default();
+        let TokenScheme::Simple(ref current_token_scheme) = current_state.token_scheme;
+
+        // No native tokens should be referenced after to the foundry destruction.
```
Comment (Alex6323) [inline bee-message/src/output/foundry.rs]:
```suggestion
        // No native tokens should be referenced after the foundry destruction.
```

File: bee-message/src/output/foundry.rs (hunk 2 above)
Comment (thibault-martinez) [reply bee-message/src/output/foundry.rs:371]: No, why ?

File: bee-message/src/output/foundry.rs [hunk 5]
```diff
@@ -383,17 +397,73 @@ impl StateTransitionVerifier for FoundryOutput {
...
+        if context.output_native_tokens.get(&token_id).is_some() {
```
Comment (Alex6323) [inline bee-message/src/output/foundry.rs]: no `contains(&token_id)` API available?

File: bee-message/src/output/foundry.rs (hunk 1 above)
Comment (thibault-martinez) [reply bee-message/src/output/foundry.rs]: Changed

File: bee-message/src/output/foundry.rs (hunk 5 above)
Comment (thibault-martinez) [reply bee-message/src/output/foundry.rs]: Changed

File: bee-message/src/output/foundry.rs (hunk 2 above)
Comment (Alex6323) [reply bee-message/src/output/foundry.rs:371]: Just trying to understand the logic. I don't see the exact types unfortunately but I guess the default would be a `0u64` or something. Line 379 would mean for the default that no tokens are minted at all during the creation which seemed strange to me, but I am probably having a misunderstanding here.

Comment (thibault-martinez) [reply bee-message/src/output/foundry.rs:371]: You can create the foundry and mint tokens at a later stage if you want to

Comment (Alex6323) [reply bee-message/src/output/foundry.rs:371]: alright, thanks for the explanation.

File: bee-message/src/output/foundry.rs (hunk 3 above)
Comment (thibault-martinez) [reply bee-message/src/output/foundry.rs]: Done

**Review by Alex6323 — APPROVED** [review]

**Review by grtlr — APPROVED** [review]

**Review by samuel-rufi — APPROVED** [review]

**Pull request merged**